akka-persistence/src/main/scala/akka/persistence/Plugins.scala

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (C) 2018 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.persistence
+
+import java.util.concurrent.atomic.AtomicReference
+
+import akka.actor.{ExtendedActorSystem, Extension, ExtensionId}
+import akka.annotation.InternalApi
+import akka.event.Logging
+import akka.persistence.Plugin.PluginHolder
+import com.typesafe.config.Config
+
+import scala.annotation.tailrec
+import scala.reflect.ClassTag
+import scala.util.Failure
+
+/**
+  * INTERNAL API
+  */
+@InternalApi
+private[akka] object Plugin {
+  private[persistence] case class PluginHolder[ScalaDsl, JavaDsl](
+    scaladslPlugin: ScalaDsl, javadslPlugin: JavaDsl)
+    extends Extension
+}
+
+/**
+  * INTERNAL API
+  */
+@InternalApi
+private[akka] trait PluginProvider[T, ScalaDsl, JavaDsl] {
```

Comment:
We use different names in various plugins. Went with a type class approach as not to change the public traits as java gets confused if we had a default impl of these methods that referes to the abstract method.